src/IRMatch.cpp

```diff
@@ -283,4 +283,126 @@ bool expr_match(Expr pattern, Expr expr, map<string, Expr> &matches) {
     }
 }
 
+namespace IRMatcher {
+
+HALIDE_ALWAYS_INLINE
+bool equal_helper(const Expr &a, const Expr &b) {
+    return equal(*a.get(), *b.get());
+}
+
+template<typename Op>
```

Comment:
Would it work to add a 'bool commutative' template parameter, and checking the commutation of the test? A lot of the rules in the select visitor below look like they're just permutations of the same rule due to commutativity...

Reply:
I have concluded that it's better to list the commutations rather than hide potentially exponential behavior. A lot of the time canonicalizations make some of the commutations impossible. It's also less applicable than you would think. Consider the following set of rules, which are all just commutations of a single underlying rule:

```
                  rewrite((x + y) + w < x + z, y + w < z) ||
                  rewrite((y + x) + w < x + z, y + w < z) ||
                  rewrite(w + (x + y) < x + z, y + w < z) ||
                  rewrite(w + (y + x) < x + z, y + w < z) ||
                  rewrite((x + y) + w < z + x, y + w < z) ||
                  rewrite((y + x) + w < z + x, y + w < z) ||
                  rewrite(w + (x + y) < z + x, y + w < z) ||
                  rewrite(w + (y + x) < z + x, y + w < z) ||
```

With the existing of a commutative add matcher called c_add, you could imagine compressing this to at least the following, and possible further:

```
                  rewrite((x + y) + w < c_add(z, x), y + w < z) ||
                  rewrite((y + x) + w < c_add(z, x), y + w < z) ||
                  rewrite(w + (x + y) < c_add(z, x), y + w < z) ||
                  rewrite(w + (y + x) < c_add(z, x), y + w < z) ||
```

But this doesn't really work very cleanly, because z is unbound, so if you try to match it to x it will succeed, and then you discover matching the other arg against x doesn't work, so you need to unbind z and roll back the state. Even if I implemented that, no further compressions of that set are actually possible, because the vars always match on their first use, when they are unbound.